File: plugins/harness/skills/harness-install/assets/maven/harness-maven-plugin/src/main/java/ai/harness/maven/ForbidScaffoldLeaksRule.java
package ai.harness.maven;

import tools.jackson.databind.JsonNode;
import org.apache.maven.plugin.MojoExecutionException;
import java.nio.file.Path;
import java.util.Collection;
import java.util.regex.Pattern;
import java.util.stream.Stream;

/**
 * Rule that forbids scaffold template patterns in active assets.
 */
public enum ForbidScaffoldLeaksRule implements HarnessCheckRule {
    INSTANCE;
    private static final String CATEGORY = "forbidScaffoldLeaks";

    @Override
    public boolean applies(JsonNode manifest) {
        return HarnessCheckHelper.applies(manifest, CATEGORY);
    }

    @Override
    public Collection<Finding> validate(Path root, JsonNode manifest) throws MojoExecutionException {
        JsonNode catNode = manifest.get(CATEGORY);
        JsonNode scope = catNode.get("parameters").get("scope");
        List<String> bases = HarnessCheckHelper.extractPaths(scope.get("bases"));
        List<String> excluded = HarnessCheckHelper.extractPaths(scope.get("excludedSubtrees"));
        List<String> extensions = HarnessCheckHelper.extractPaths(scope.get("extensions"));
        JsonNode patternsNode = catNode.get("parameters").get("patterns");
        String severity = HarnessCheckHelper.getSeverity(manifest, CATEGORY);
        List<Path> excludedPaths = excluded.stream().map(root::resolve).toList();
        String extRegex = extensions.isEmpty() ? ".*" : ".*\\.(" + String.join("|", extensions.stream().map(Pattern::quote).toList()) + ")$";
        Pattern extPattern = Pattern.compile(extRegex);
        return bases.stream()
                .flatMap(baseName -> {
                    try {
                        Path base = root.resolve(baseName);
                        return HarnessCheckHelper.safeFileOrWalk(root, base).stream()
                                .filter(file -> !excludedPaths.stream().anyMatch(file::startsWith))
                                .filter(file -> extPattern.matcher(file.toString()).matches())
                                .flatMap(file -> checkLeaks(root, file, patternsNode, severity).stream());
                    } catch (MojoExecutionException e) {
                        return Stream.empty();
                    }
                })
                .toList();
    }

    private List<Finding> checkLeaks(Path root, Path file, JsonNode patternsNode, String severity) throws MojoExecutionException {
        String text = HarnessCheckHelper.readFile(root, file);
        return Stream.of(patternsNode.spliterator(), false)
                .filter(p -> Pattern.compile(p.get("pattern").asText()).matcher(text).find())
                .map(p -> new Finding(severity, CATEGORY, p.get("label").asText() + " in active asset: " + root.relativize(file)))
                .toList();
    }
}

File: plugins/harness/skills/harness-install/assets/gradle/buildSrc/src/main/kotlin/ai/harness/gradle/ForbidScaffoldLeaksRule.kt
package ai.harness.gradle

import kotlinx.serialization.json.JsonObject
import java.nio.file.Path
import kotlin.io.path.div
import kotlin.io.path.extension
import kotlin.io.path.isRegularFile
import kotlin.io.path.readText
import kotlin.io.path.relativeTo
import java.io.File

/**
 * Rule that forbids scaffold placeholder patterns in active assets.
 */
object ForbidScaffoldLeaksRule : HarnessCheckRule {
	override fun applies(manifest: JsonObject): Boolean {
		val category = "forbidScaffoldLeaks"
		val catObj = manifest[category]?.jsonObject ?: return false
		val enabled = catObj["enabled"]?.jsonPrimitive?.contentOrNull?.toBoolean() ?: true
		return enabled
	}

	override fun validate(manifest: JsonObject, root: Path, psiResults: HarnessPsiResults?): Collection<Finding> = buildSet {
		val category = "forbidScaffoldLeaks"
		val severity = HarnessCheck.Companion.severityOf(manifest, category)
		val catObj = manifest[category]?.jsonObject ?: return@buildSet
		val parametersObj = catObj["parameters"]?.jsonObject ?: return@buildSet
		val messagesObj = catObj["messages"]?.jsonObject ?: return@buildSet
		val scopeObj = parametersObj["scope"]?.jsonObject ?: return@buildSet
		val bases = HarnessCheck.Companion.stringArrayFrom(scopeObj, "bases")
		val excludedSubtrees = HarnessCheck.Companion.stringArrayFrom(scopeObj, "excludedSubtrees")
		val extensions = HarnessCheck.Companion.stringArrayFrom(scopeObj, "extensions")
		val patterns = parametersObj["patterns"]?.jsonArray?.mapNotNull { patternElem ->
			val obj = patternElem.jsonObject
			val pattern = HarnessCheck.Companion.stringFrom(obj, "pattern")
			val label = HarnessCheck.Companion.stringFrom(obj, "label")
			if (pattern.isEmpty() || label.isEmpty()) null else pattern to label
		} ?: emptyList()

		val excludedPaths = excludedSubtrees.map { root / it }
		val regexes = patterns.mapNotNull { (pattern, label) ->
			try {
				pattern.toRegex() to label
			} catch (_: Exception) {
				null
			}
		}

		return buildSet<Finding> {
			bases.forEach { basePath ->
				val base = root / basePath
				val (files, _) = HarnessCheck.Companion.walkSafe(root, base)
				files.forEach { file ->
					val ext = file.extension
					if (ext in extensions && excludedPaths.none { file.toString().startsWith(it.toString()) }) {
						val text = file.readText()
						regexes.forEach { (regex, label) ->
							if (regex.containsMatchIn(text)) {
								val msg = HarnessCheck.Companion.stringFrom(messagesObj, "default").takeIf { it.isNotEmpty() } ?: "$label in active asset: ${file.relativeTo(root.toFile())}"
								add(Finding(severity, category, msg))
							}
						}
					}
				}
			}
		}.toList()
	}
}
